iroh/src/downloader.rs

```diff
@@ -0,0 +1,979 @@
+    fn handle_cancel_download(&mut self, id: Id, kind: DownloadKind) {
+        if let Entry::Occupied(mut occupied_entry) = self.current_requests.entry(kind.clone()) {
+            // remove the intent from the associated request
+            let intents = &mut occupied_entry.get_mut().intents;
+            intents.remove(&id);
+            // if this was the last intent associated with the request cancel it
+            if intents.is_empty() {
+                occupied_entry.remove().cancellation.cancel();
+            }
+        } else if let Entry::Occupied(mut occupied_entry) = self.scheduled_requests.entry(kind) {
+            // remove the intent from the associated request
+            let intents = &mut occupied_entry.get_mut().intents;
+            intents.remove(&id);
+            // if this was the last intent associated with the request remove it from the schedule
+            // queue
+            if intents.is_empty() {
+                let delay_key = occupied_entry.remove().delay_key;
+                self.scheduled_request_queue.remove(&delay_key);
+            }
+        }
+    }
+
+    /// Handle a [`Message::PeersHave`].
+    fn handle_peers_have(&mut self, hash: Hash, peers: Vec<PublicKey>) {
+        // check if this still needed
+        if self.is_needed(hash) {
+            self.providers.add_peers(hash, &peers);
+        }
+    }
+
+    /// Checks if this hash is needed.
+    fn is_needed(&self, hash: Hash) -> bool {
+        let as_blob = DownloadKind::Blob { hash };
+        let as_collection = DownloadKind::Collection { hash };
+        self.current_requests.contains_key(&as_blob)
+            || self.scheduled_requests.contains_key(&as_blob)
+            || self.current_requests.contains_key(&as_collection)
+            || self.scheduled_requests.contains_key(&as_collection)
+    }
+
+    /// Handle receiving a new connection.
+    fn on_connection_ready(&mut self, peer: PublicKey, result: anyhow::Result<D::Connection>) {
+        match result {
+            Ok(connection) => {
+                trace!(%peer, "connected to peer");
+                let drop_key = self.goodbye_peer_queue.insert(peer, IDLE_PEER_TIMEOUT);
+                self.peers
+                    .insert(peer, ConnectionInfo::new_idle(connection, drop_key));
+            }
+            Err(err) => {
+                debug!(%peer, %err, "connection to peer failed")
+            }
+        }
+    }
+
+    fn on_download_completed(&mut self, kind: DownloadKind, result: Result<(), FailureAction>) {
+        // first remove the request
+        let info = self
+            .current_requests
+            .remove(&kind)
+            .expect("request was active");
+
+        // update the active requests for this peer
+        let ActiveRequestInfo {
+            intents,
+            peer,
+            mut remaining_retries,
+            ..
+        } = info;
+
+        let peer_info = self.peers.get_mut(&peer).expect("peer is connected");
```

Comment:
Is this guranteed to be true? couldn't we have disconnected in the meantime?

Reply:
maybe the connection was lost, but the peer should only be removed from the mapping after being idle, so having active quests = 0. If we get here the peer has active requests > 0

I will check in any case

Reply:
I changed the expect message to better reflect this logic in [aba53f2e](https://github.com/n0-computer/iroh/pull/1420/commits/aba53f2e60dd5d79f7588a9ea152237b847e0548#diff-d4ea9132b751f09c6923e77e3e0fa5a8dfaa3971da42cb12d99837f06f900505L712)